### PR TITLE
defer plugin contributions (preferences and tasks)

### LINKED ISSUE #7672: [plugin] fetching tasks is slow for big multi-root workspaces
`validateSchema` should be computed once and cached

but I am not sure that we need it at all, does VS Code validate task configs? 

We removed such logic already for preferences, since it is slow and VS Code does not care about.

cc @elaihau @RomanNikitenko

### LINKED ISSUE #7673: register plugin contributions is slow in big workspaces
We have some very big workspaces like multi-root with over 12 roots with many files plus terminals building it. UI gets completely unresponsive for 5s.

There are 2 reasons mostly:
- https://github.com/eclipse-theia/theia/issues/7672
- and preference reconciliation

I will look into preferences.

### PR DESCRIPTION
**Pull request by akosyakov**
#### What it does

- fix #7673: with many VS Code extensions repetitive events are happening like update editor configurations and task schemas. This PR makes sure that such event defer changes in the same tick and perform them together.
- fix #7672: simplify `$fetchTasks` by removing filtering duplicate tasks. Testing showed that VS Code does not do it (see https://github.com/eclipse-theia/theia/issues/7672#issuecomment-620464015) and even after removing our implementation fails to return duplicate tasks. That's apparently an issue how the task extension treats the task identity: https://github.com/eclipse-theia/theia/issues/7681#issuecomment-620482611

TODO:
- [x] fix preference tests to check that there is not changes if schema set and unset in the same tick and there are if in different ticks
- [x] make sure that update/replace configuration calls from the extension host process are not resolved before preference changes are propagated to the extension host

#### How to test

- check out https://github.com/akosyakov/jupyterlab/tree/ak/fetch_tasks
```bash
git clone https://github.com/akosyakov/jupyterlab.git \
cd jupyterlab/ \
git checkout ak/fetch_tasks
```
- start Theia for [jlab.theia-workspace](https://github.com/akosyakov/jupyterlab/blob/ak/fetch_tasks/jlab.theia-workspace) workspace
```bash
yarn start ../../../jupyterlab/jlab.theia-workspace
```
- open README.md as an editor and for preview
- enable CPU profiler and reload the page
- check that that registering plugin contributions is fast, not like here https://github.com/eclipse-theia/theia/issues/7673#issuecomment-620007593

Testing task fetching:
- install https://github.com/akosyakov/vscode-fetch-tasks-test
```bash
cd plugins/ \
git clone https://github.com/akosyakov/vscode-fetch-tasks-test.git \
cd vscode-fetch-tasks-test/ \
yarn && yarn compile
```
- restart Theia to load the plugin
- use `Hello World` to see all found tasks in output view
- you can also test according to https://github.com/eclipse-theia/theia/pull/6058 and https://github.com/eclipse-theia/theia/pull/5975

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/core/src/browser/preferences/preference-provider.ts [hunk 1]
```diff
@@ -58,22 +59,34 @@ export abstract class PreferenceProvider implements Disposable {
...
             for (const change of changes) {
-                prefChanges[change.preferenceName] = change;
+                this.deferredChanges[change.preferenceName] = change;
             }
-            this.onDidPreferencesChangedEmitter.fire(prefChanges);
         } else {
-            this.onDidPreferencesChangedEmitter.fire(changes);
+            this.deferredChanges = Object.assign(this.deferredChanges, changes);
         }
+        this.fireDidPreferencesChanged();
```
Comment (akosyakov) [inline packages/core/src/browser/preferences/preference-provider.ts:79]: use `setImmediate` instead to make sure that events are fired at the end of the same tick, not in next tick

File: packages/task/src/browser/task-schema-updater.ts [hunk 2]
```diff
@@ -70,12 +72,15 @@ export class TaskSchemaUpdater {
         this.taskDefinitionRegistry.onDidUnregisterTaskDefinition(() => this.updateSupportedTaskTypes());
     }
 
-    update(): void {
+    readonly update = debounce(() => this.doUpdate(), 0);
```
Comment (akosyakov) [inline packages/task/src/browser/task-schema-updater.ts:75]: use `setImmediate` instead to make sure that events are fired at the end of the same tick, not in next tick

Comment (akosyakov) [reply packages/task/src/browser/task-schema-updater.ts:75]: not really matter here

File: packages/core/src/browser/preferences/preference-provider.ts (hunk 1 above)
Comment (akosyakov) [reply packages/core/src/browser/preferences/preference-provider.ts:79]: `setImmediate` is not standard API in the browser, so no

File: packages/core/src/browser/preferences/preference-service.spec.ts [hunk 3]
```diff
@@ -185,12 +153,56 @@ describe('Preference Service', () => {
...
+
+        assert.deepStrictEqual([], events.map(e => ({
+            preferenceName: e.preferenceName,
+            newValue: e.newValue,
+            oldValue: e.oldValue
+        })), 'events after unset in the same tick');
+        assert.strictEqual(prefService.get('editor.insertSpaces'), undefined, 'get after');
+        assert.strictEqual(prefService.get('[go].editor.insertSpaces'), undefined, 'get after overridden');
+
+        await new Promise(resolve => setTimeout(resolve, 50));
```
Comment (akosyakov) [inline packages/core/src/browser/preferences/preference-service.spec.ts:194]: replace with `assert.strictEqual(await prefSchema.pendingChanges, false);`

**Comment by RomanNikitenko** [comment]
@akosyakov 
> simplify $fetchTasks by removing filtering duplicate tasks

I played with your changes this morning and I think the filtering was related to customized tasks.
The difference is:
- before your changes it returned `configured` and `detected` tasks
- with your changes it returns `configured`, `detected` and `customized` tasks

To reproduce it with your plugin I modified a little your plugin to see whole configuration, not only `name` and `source`, then:
- use `Hello World` to see all found tasks in output view
- pay attention on number of tasks
- go to `Terminal` -> `Configure task` and configure a few tasks   
- use `Hello World` to see all found tasks in output view
- pay attention on number of tasks

I don't know how vs code behaves for this case.
Sorry for the delay, maybe this comment will be useful for https://github.com/eclipse-theia/theia/issues/7681.

**Pull request merged**